ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/EventsTableTreeEvents.scala

```diff
@@ -175,7 +175,26 @@ private[events] trait EventsTableTreeEvents { this: EventsTable =>
       previousOffsetWhereClauseValues(startExclusive, previousEventNodeIndex)
     val witnessesWhereClause =
       sqlFunctions.arrayIntersectionWhereClause("tree_event_witnesses", requestingParty)
-    SQL"select #$selectColumns, array[$requestingParty] as event_witnesses, case when submitter = $requestingParty then command_id else '' end as command_id from participant_events where (event_offset > $startExclusive or (event_offset = $prevOffset and node_index > $prevNodeIndex)) and event_offset <= $endInclusive and #$witnessesWhereClause order by (#$orderByColumns) limit $pageSize"
+    SQL"""select #$selectColumns, array[$requestingParty] as event_witnesses, case when submitter = $requestingParty then command_id else '' end as command_id
+         from participant_events,
+          (select min(row_id) row_id from (
+            select min(row_id) row_id
+            from participant_events
+            where (event_offset > $startExclusive)
+              and event_offset <= $endInclusive
+              and #$witnessesWhereClause
+           union all
+            -- this sub-query needs the index participant_event_event_offset_idx to run fast
+            select min(row_id) row_id
+            from participant_events
+            where
+              (event_offset = $prevOffset and node_index > $prevNodeIndex)
+              and #$witnessesWhereClause
+            ) rows) starting_row
+            where
+            participant_events.row_id >= starting_row.row_id
+            and participant_events.row_id < starting_row.row_id + $pageSize
+            and event_offset <= $endInclusive and #$witnessesWhereClause order by (#$orderByColumns) """
```

Comment:
```suggestion
            and event_offset <= $endInclusive and #$witnessesWhereClause order by (row_id) limit $pageSize"""
```

The benefit comes from the synthetic `row_id` and a row_id-major index, not the removal of the `limit` per se.

Reply:
```
sandbox=# explain analyze select * from participant_events where row_id >= 903006 order by row_id LIMIT 1000;
                                                                                           QUERY PLAN                                                                                           
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.43..1889.50 rows=1000 width=1694) (actual time=0.039..1.282 rows=1000 loops=1)
   ->  Index Scan using participant_events_row_id_event_offset_node_index_idx on participant_events  (cost=0.43..813918.41 rows=430855 width=1694) (actual time=0.037..1.183 rows=1000 loops=1)
         Index Cond: (row_id >= 903006)
 Planning Time: 0.195 ms
 Execution Time: 1.348 ms
```

Reply:
It's possible that reference to `event_offset` and the contents of `witnessesWhereClause` may complicate this; we will have to do more complex trials. Regardless, the presence of `witnessesWhereClause` is another reason aside from sequence gaps that we can't do row_id arithmetic exactly.

Reply:
A further demonstration:

```sql
explain analyze select * from participant_events
where row_id >= 128 and flat_event_witnesses && array['00001']::varchar[]
order by row_id LIMIT 1000;
                                                                               QUERY PLAN                                                                               
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.42..700.68 rows=1000 width=1816) (actual time=0.026..1.243 rows=1000 loops=1)
   ->  Index Scan using participant_events_row_id_idx on participant_events  (cost=0.42..265997.37 rows=379854 width=1816) (actual time=0.025..1.097 rows=1000 loops=1)
         Index Cond: (row_id >= 128)
         Filter: (flat_event_witnesses && '{00001}'::character varying[])
         Rows Removed by Filter: 33
 Planning Time: 1.070 ms
 Execution Time: 1.343 ms
```

We can see that the index _precludes_ an actual sort; however, the linear filter for visibility entails an unbounded scan of the (presorted by index) events table looking for elements passing the predicate (i.e. witnesses, and also template in the non-wildcard case).

Because we can return fewer than pageSize events, must return at least one element, and _if scanning > pageSize elements means we should stop as soon as possible_, it seems the most straightforward way to meet these constraints would be to

1. Perform an initial scan with row_id arithmetic. If that set is non-empty, return it, however large it is.
2. If not, perform a scan with limit 1; just return a result set with that element.

(I think the implication of the second is potentially degenerate behavior, but meets the constraints we have discussed so far, which may be worth reconsidering. However, it does have a lot of good qualities, too, for practical client interaction with the transaction stream.)

The first of these looks like

```sql
explain analyze select * from participant_events
where row_id >= 128 and row_id < 1128 and flat_event_witnesses && array['00001']::varchar[]
order by row_id;
                                                                         QUERY PLAN                                                                         
------------------------------------------------------------------------------------------------------------------------------------------------------------
 Index Scan using participant_events_row_id_idx on participant_events  (cost=0.42..1251.41 rows=858 width=1816) (actual time=0.060..2.628 rows=968 loops=1)
   Index Cond: ((row_id >= 128) AND (row_id < 1128))
   Filter: (flat_event_witnesses && '{00001}'::character varying[])
   Rows Removed by Filter: 32
 Planning Time: 6.333 ms
 Execution Time: 2.764 ms
```

Presuming that failed, we would start at the former row_id upper bound and keep looking until we find 1 element. This can be costly, because there is no rational way to apply the key constraint with which we seek to get further gains with this change--the row_id upper bound--especially in the case where the predicate is difficult to satisfy, but seems to be the minimum amount of work to satisfy the API invariants.

Compare a probably-true predicate to a probably-false predicate's plan:

```sql
explain analyze select * from participant_events
where row_id >= 1128 and flat_event_witnesses && array['11111']::varchar[]
order by row_id limit 1;
                                                                             QUERY PLAN                                                                             
--------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.42..6.46 rows=1 width=1816) (actual time=0.050..0.050 rows=1 loops=1)
   ->  Index Scan using participant_events_row_id_idx on participant_events  (cost=0.42..265700.98 rows=44020 width=1816) (actual time=0.049..0.049 rows=1 loops=1)
         Index Cond: (row_id >= 1128)
         Filter: (flat_event_witnesses && '{11111}'::character varying[])
         Rows Removed by Filter: 2
 Planning Time: 0.162 ms
 Execution Time: 0.071 ms


explain analyze select * from participant_events where row_id >= 1128
and flat_event_witnesses && array['00010']::varchar[]
order by row_id limit 1;
                                                                            QUERY PLAN                                                                            
------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Limit  (cost=143.75..143.75 rows=1 width=1816) (actual time=0.104..0.104 rows=1 loops=1)
   ->  Sort  (cost=143.75..143.81 rows=23 width=1816) (actual time=0.103..0.103 rows=1 loops=1)
         Sort Key: row_id
         Sort Method: top-N heapsort  Memory: 26kB
         ->  Bitmap Heap Scan on participant_events  (cost=48.18..143.63 rows=23 width=1816) (actual time=0.087..0.091 rows=4 loops=1)
               Recheck Cond: (flat_event_witnesses && '{00010}'::character varying[])
               Filter: (row_id >= 1128)
               Rows Removed by Filter: 1
               Heap Blocks: exact=5
               ->  Bitmap Index Scan on participant_events_flat_event_witnesses_idx  (cost=0.00..48.18 rows=24 width=0) (actual time=0.076..0.076 rows=5 loops=1)
                     Index Cond: (flat_event_witnesses && '{00010}'::character varying[])
 Planning Time: 0.175 ms
 Execution Time: 8.048 ms
```